iree/hal/local/loaders/wasm_module_loader.c

```diff
@@ -0,0 +1,552 @@
+    wasm_push_constants[i] = dispatch_state->push_constants[i];
+  }
+  // TODO(scotttodd): fix:
+  //   '=': 'const uint32_t *' differs in levels of indirection from 'uint32_t'
+  wasm_dispatch_state->push_constants = wasm_buffer_push_constants;
+  // Bindings.
+  wasm_dispatch_state->binding_count = dispatch_state->binding_count;
+  //   - binding_ptrs
+  void* native_buffer_binding_ptrs = NULL;
+  uint32_t wasm_buffer_binding_ptrs = wasm_runtime_module_malloc(
+      executable->wasm_module_inst,
+      dispatch_state->binding_count * sizeof(int32_t*),
+      &native_buffer_binding_ptrs);
+  if (IREE_UNLIKELY(wasm_buffer_binding_ptrs == 0)) {
+    return iree_make_status(IREE_STATUS_RESOURCE_EXHAUSTED, "malloc failed");
+  }
+  int32_t* wasm_binding_ptrs = (int32_t*)native_buffer_binding_ptrs;
+  // HACK: remember the last native buffer (usually the single output buffer)
+  //   This is broken in the general case, but lets us memcpy results back
+  //   from WASM to the host, without extra bookkeeping.
+  //   Revisit when using a real allocation scheme.
+  void* latest_native_buffer_binding = NULL;
+  for (int i = 0; i < dispatch_state->binding_count; ++i) {
+    void* native_buffer_binding = NULL;
+    uint32_t wasm_buffer_binding = wasm_runtime_module_malloc(
+        executable->wasm_module_inst, dispatch_state->binding_lengths[i],
+        &native_buffer_binding);
+    if (IREE_UNLIKELY(wasm_buffer_binding == 0)) {
+      return iree_make_status(IREE_STATUS_RESOURCE_EXHAUSTED, "malloc failed");
+    }
+    memcpy(native_buffer_binding, dispatch_state->binding_ptrs[i],
+           dispatch_state->binding_lengths[i]);
+    latest_native_buffer_binding = native_buffer_binding;
+    wasm_binding_ptrs[i] = wasm_buffer_binding;
+  }
+  // TODO(scotttodd): fix
+  //   '=': 'void *const *' differs in levels of indirection from 'uint32_t'
+  wasm_dispatch_state->binding_ptrs = wasm_buffer_binding_ptrs;
+  //   - binding_lengths
+  void* native_buffer_binding_lengths = NULL;
+  uint32_t wasm_buffer_binding_lengths = wasm_runtime_module_malloc(
+      executable->wasm_module_inst,
+      dispatch_state->binding_count * sizeof(size_t*),
+      &native_buffer_binding_lengths);
+  if (IREE_UNLIKELY(wasm_buffer_binding_lengths == 0)) {
+    return iree_make_status(IREE_STATUS_RESOURCE_EXHAUSTED, "malloc failed");
+  }
+  int32_t* wasm_binding_lengths = (int32_t*)native_buffer_binding_lengths;
+  for (int i = 0; i < dispatch_state->binding_count; ++i) {
+    wasm_binding_lengths[i] = dispatch_state->binding_lengths[i];
+  }
+  // TODO(scotttodd): fix
+  //   '=': 'const size_t *' differs in levels of indirection from 'uint32_t'
+  wasm_dispatch_state->binding_lengths = wasm_buffer_binding_lengths;
+  // Imports.
+  // TODO(scotttodd): implement import table if needed? fail if not-NULL?
+  wasm_dispatch_state->imports = NULL;
+
+  // Clone workgroup_id.
+  void* native_buffer_workgroup_id = NULL;
```

Comment:
these too can be part of the single allocation for the dispatch state (following it in memory), with the pointers sliced out the same way.